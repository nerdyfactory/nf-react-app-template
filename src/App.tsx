import React, { useState, useReducer, useEffect } from 'react';
import { ThemeProvider } from '@material-ui/styles/';
import { BrowserRouter } from 'react-router-dom';
import { Toaster } from 'react-hot-toast';

import { PrivateNav, PublicNav } from 'navigations';
import useBootstrap from 'hooks/useBootstrap';
import { AppContext, appReducer, INITIAL_APP_STATE } from 'contexts/AppContext';
import { AuthActionTypes, AuthContext, authReducer, INITIAL_AUTH_STATE } from 'contexts';
import { Loading } from 'components';
import './App.css';
import { USER_TOKEN } from 'constants/utility';
import { getIntialState } from 'services/utils/persist-state';

function App() {
  const [authCtx, authDispatcher] = useReducer(authReducer, INITIAL_AUTH_STATE);
  const [appCtx, appDispatcher] = useReducer(appReducer, INITIAL_APP_STATE);
  const [theme] = useBootstrap();
  const [isLoaded, setIsLoaded] = useState(false);

  useEffect(() => {
    const token = getIntialState(USER_TOKEN);
    if (token) {
      authDispatcher({ type: AuthActionTypes.SET_TOKEN, payload: token });
    }
    setIsLoaded(true);
  }, []);

  if (!isLoaded) return <Loading />;

  return (
    <ThemeProvider theme={theme}>
      <BrowserRouter basename="/">
        <AuthContext.Provider value={{ ...authCtx, dispatch: authDispatcher }}>
          <AppContext.Provider value={{ ...appCtx, dispatch: appDispatcher }}>
            {authCtx.token ? <PrivateNav /> : <PublicNav />}
            <Toaster
              position="top-right"
              toastOptions={{
                style: {
                  margin: '40px',
                  background: '#363636',
                  color: '#fff',
                  zIndex: 99999999,
                },
              }}
            />
          </AppContext.Provider>
        </AuthContext.Provider>
      </BrowserRouter>
    </ThemeProvider>
  );
}

export default App;
