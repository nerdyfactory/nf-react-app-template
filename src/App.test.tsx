import React from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import App from './App';

describe('App', () => {
  it('renders app', () => {
    const { container } = render(<App />);
    expect(container).toMatchSnapshot();
  });

  describe('Login', () => {
    it('moves to logout when login', async () => {
      render(<App />);
      fireEvent.click(screen.getByText('Login'));
      await waitFor(() => screen.getByRole('button'));
      expect(screen.getByRole('button')).toHaveTextContent('Logout');
    });
  });

  describe('Logout', () => {
    const loginAfterRender = () => {
      render(<App />);
      fireEvent.click(screen.getByText('Login'));
    };

    it('moves to login screen', async () => {
      loginAfterRender();
      fireEvent.click(screen.getByText('Logout'));
      await waitFor(() => screen.getByRole('button'));
      expect(screen.getByRole('button')).toHaveTextContent('Login');
    });
  });
});
