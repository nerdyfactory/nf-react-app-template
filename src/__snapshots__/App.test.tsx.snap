// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders app 1`] = `
<div>
  <div
    class="App"
  >
    <div>
      <button>
        Login
      </button>
    </div>
  </div>
</div>
`;
